Adds support for controlling Growatt MIN/TLX inverters through number platform and entities

johanzander opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds support for controlling MIN/TLX inverters through number entities when using the V1 API (token authentication).

This enables users to adjust battery charging/discharging parameters directly from Home Assistant:

Charge power rate, 
Charge stop SOC, 
Discharge power rate, 
Discharge stop SOC 
(all as percentages)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR is replaces https://github.com/home-assistant/core/pull/153468 which will be closed. 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/41285

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

johanzander committed:
Add number platform support for Growatt with entity creation and API integration

johanzander committed:
Merge branch 'dev' of https://github.com/home-assistant/core into growatt-number-platform

johanzander committed:
Merge branch 'dev' into growatt-number-platform

davidrapan reviewed (changes requested):
`_rate` suffix at `charge_power_rate` and `discharge_power_rate` doesn't belong here.

joostlek commented on homeassistant/components/growatt_server/number.py:
```diff
@@ -0,0 +1,169 @@
+
+    async def async_set_native_value(self, value: float) -> None:
+        """Set the value of the number."""
+        try:
+            # Use write_key if specified, otherwise fall back to api_key
+            parameter_id = (
+                self.entity_description.write_key or self.entity_description.api_key
+            )
+
+            # Use V1 API to write parameter
+            await self.hass.async_add_executor_job(
+                self.coordinator.api.min_write_parameter,
+                self.coordinator.device_id,
+                parameter_id,
+                int(value),
+            )
+
+            _LOGGER.debug(
+                "Set parameter %s to %s",
+                parameter_id,
+                value,
+            )
+
+            # If no exception was raised, the write was successful
+            # Update the value in coordinator data
+            self.coordinator.data[self.entity_description.api_key] = int(value)
+            self.async_write_ha_state()
+
+        except GrowattV1ApiError as e:
+            _LOGGER.error("Error while setting parameter: %s", e)
+            raise HomeAssistantError(f"Error while setting parameter: {e}") from e
```
Only have things in the try block that can raise

joostlek commented on homeassistant/components/growatt_server/number.py:
```diff
@@ -0,0 +1,169 @@
+            raise HomeAssistantError(f"Error while setting parameter: {e}") from e
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: GrowattConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up Growatt number entities."""
+    runtime_data = entry.runtime_data
+
+    entities: list[GrowattNumber] = []
+
+    # Add number entities for each MIN device (only supported with V1 API)
+    for device_coordinator in runtime_data.devices.values():
+        if (
+            device_coordinator.device_type == "min"
+            and device_coordinator.api_version == "v1"
+        ):
+            entities.extend(
+                GrowattNumber(
+                    coordinator=device_coordinator,
+                    description=description,
+                )
+                for description in MIN_NUMBER_TYPES
+            )
+
+    async_add_entities(entities)
```
Please place this above the entity class and underneath the descriptions

johanzander commented:
@davidrapan  - correct.

what do you think of the following instead:
  1. key="battery_charge_power_limit"     # Power limit for charging
  2. key="battery_charge_stop_soc"        # Stop charging at this  SOC
  3. key="battery_discharge_power_limit"  # Power limit for discharging  
  4. key="battery_discharge_stop_soc"     # Stop discharging at this SOC
  
I think battery_charge_power is too ambiguous as this is percentage and not power (there is a charge power sensor as well)

davidrapan commented:
This is much better. 👍 And I would use it for all added number entities (e.g. `battery_charge_soc_limit`), but it's more like nitpicking than a request for change, hah.

> there is a charge power sensor as well

_Well, if it's sensor entity then it can share the same `entity_id` w/ number entity, just for a sidenote._

johanzander committed:
Fix review comments: Rename Growatt number entity keys and corresponding strings for clarity + only have raise stuff in try block, increased unit test coverage

johanzander committed:
Merge branch 'dev' of https://github.com/home-assistant/core into growatt-number-platform

johanzander committed:
Merge branch 'growatt-number-platform' of https://github.com/johanzander/core into growatt-number-platform

johanzander committed:
Merge branch 'dev' into growatt-number-platform

johanzander committed:
Refactor Growatt server test fixtures and enhance number entity tests for V1 API integration

johanzander committed:
Merge branch 'dev' of https://github.com/home-assistant/core into growatt-number-platform

johanzander committed:
Merge branch 'growatt-number-platform' of https://github.com/johanzander/core into growatt-number-platform

johanzander committed:
adress review comments

johanzander committed:
Merge branch 'dev' of https://github.com/home-assistant/core into growatt-number-platform

johanzander committed:
Merge branch 'dev' into growatt-number-platform

joostlek reviewed (changes requested):
There's a merge conflict

johanzander committed:
Include switch platform support in number platform branch

Updated growatt_server integration to support both NUMBER and SWITCH platforms by manually merging changes from upstream/dev (where switch platform was already merged).

Changes:
- const.py: Added Platform.SWITCH to PLATFORMS list
- strings.json: Added switch entity translations for ac_charge
- conftest.py: Added acChargeEnable to mock data and updated comments to mention both switch and number platforms

johanzander committed:
Merge branch 'growatt-number-platform' of https://github.com/johanzander/core into growatt-number-platform

joostlek reviewed (changes requested):
There's still a merge conflict

johanzander committed:
Merge branch 'dev' into growatt-number-platform

johanzander committed:
Merge branch 'dev' into growatt-number-platform

johanzander committed:
Merge branch 'dev' into growatt-number-platform

johanzander committed:
Merge branch 'dev' into growatt-number-platform

joostlek commented on homeassistant/components/growatt_server/number.py:
```diff
@@ -0,0 +1,169 @@
+                for description in MIN_NUMBER_TYPES
+            )
+
+    async_add_entities(entities)
```
```suggestion
    # Add number entities for each MIN device (only supported with V1 API)
    async_add_entities(
        GrowattNumber(device_coordinator, description)
        for device_coordinator in runtime_data.devices.values()
        if (
            device_coordinator.device_type == "min"
            and device_coordinator.api_version == "v1"
        )
        for description in MIN_NUMBER_TYPES
    )
```

joostlek commented on homeassistant/components/growatt_server/number.py:
```diff
@@ -0,0 +1,169 @@
+
+        # Update the value in coordinator data
+        self.coordinator.data[self.entity_description.api_key] = int_value
+        self.async_write_ha_state()
```
just as a double check, we can't ask the coordinator to refresh to get new data?

joostlek commented on homeassistant/components/growatt_server/number.py:
```diff
@@ -0,0 +1,169 @@
+            )
+        except GrowattV1ApiError as e:
+            _LOGGER.error("Error while setting parameter: %s", e)
+            raise HomeAssistantError(f"Error while setting parameter: {e}") from e
```
Let's not log and raise, let's remove the log

joostlek reviewed (changes requested):
final final things

johanzander committed:
Update homeassistant/components/growatt_server/number.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

johanzander committed:
fixes review comments

johanzander committed:
Merge branch 'dev' into growatt-number-platform
